xls/modules/rle/rle.x

```diff
@@ -0,0 +1,324 @@
+    // zero means that the previous evaluation sent all the data and
+    // we start counting from the beginning
+    prev_count: bits[COUNT_WIDTH],
+    // flag indicating that the previous symbol was the last one
+    // in the transmission
+    prev_last: bool,
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<EncInData<SYMBOL_WIDTH>> in;
+    output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev_symbol: bits[SYMBOL_WIDTH]:0,
+            prev_count: bits[COUNT_WIDTH]:0,
+            prev_last: false,
+        }
+    )}
+
+    config (
+        input_r: chan<EncInData<SYMBOL_WIDTH>> in,
+        output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = EncInData { symbol: bits[SYMBOL_WIDTH]:0, last: false };
+        let (input_tok, input) = recv_if(tok, input_r, !state.prev_last, zero_input);
+
+        let prev_symbol_valid = state.prev_count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (input.symbol != state.prev_symbol);
+        let overflow = state.prev_count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let (symbol, count, last) = if (state.prev_last) {
+            (
+                bits[SYMBOL_WIDTH]:0,
+                bits[COUNT_WIDTH]:0,
+                false
+            )
+        } else if (symbol_differ || overflow) {
+            (
+                input.symbol,
+                bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        } else {
+            (
+                input.symbol,
+                state.prev_count + bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        };
+
+        let data = EncOutData {
+            symbol: state.prev_symbol,
+            count: state.prev_count,
+            last: state.prev_last
+        };
+
+        let do_send = state.prev_last || symbol_differ || overflow;
+        let data_tok = send_if(input_tok, output_s, do_send, data);
+
+        RLEEncState {
+            prev_symbol: symbol,
+            prev_count: count,
+            prev_last: last,
+        }
+    }
+}
+
+// RLE encoder specialization for the codegen
+proc RLEEnc32 {
+
+    init {()}
+
+    config (
+        input_r: chan<EncInData<32>> in,
+        output_s: chan<EncOutData<32, 2>> out,
+    ) {
+        spawn RLEEnc<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH = u32:2;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount = bits[TEST_COUNT_WIDTH];
+type TestEncInData = EncInData<TEST_SYMBOL_WIDTH>;
+type TestEncOutData = EncOutData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+
+const TEST_SEND_LOOKUP = TestEncInData[20]:[
```

Comment:
I'd be curious to see what the test would looks like if all those payload were inlined in the test itself instead?

The code might be a little more verbose, but the flow between data send and assertion on recv'ed message would be more readable (currently you have to make an interection between the two tables to figure that out).

Reply:
@meheff also suggested that you might have more control if testing those in C++.

Reply:
It is possible, but test_proc's next function would be enormous.
It would also remove ability to quickly extend test's inputs and outputs.

As for C++ tests, on one hand,it would give as access to procs state and JIT.
On the other hand, it would remove a non-trivial test_proc example,
that uses a separate proc for stimuli and different proc for scoring DUT behavior.

Reply:
> It is possible, but test_proc's next function would be enormous.

I'm not sure if it would be that enormous, as you would be able to leverage for loops for sending multiple packets and assert each packet being received right after each group of data.

> As for C++ tests, on one hand,it would give as access to procs state and JIT 
> On the other hand, it would remove a non-trivial test_proc example,
> that uses a separate proc for stimuli and different proc for scoring DUT behavior.

Just something (C++ tests) to keep in mind for more complicated test (not necessarly this one).

Note that I do agree that having non-trivial dslx test is generally useful and that it's interesting to demonstrate using separate proc for stimuli and assertion. For that particular test, we have a tight correlation both temporal and in the data being tested between the stimuli and assertion: so I feel that we loose a bit of the this relation  with the current data-driven design.

Reply:
> So I've tried to create rather elegant solution using for over TEST_SEND_LOOKUP and recv_non_blocking to get response from RLEEnc

What happend if you interleave "for over TEST_SEND_LOOKUP + send" and "blocking recv"?

Reply:
It can be done, but it will use recv_if.
I'll have to specify when I expect to receive data from the encoder.

I must also handle double sends from RLE.
Array will by of TEST_SEND_LEN length, and store values: 0,1,2, so it can be concatenated with TEST_SEND array.

I'll push commit with such change.

Reply:
There is one more thing, I use multi process approach in other PR, and in some test cases I directly reuse `RLEEncTestSender`

Reply:
I've changed thigh test to interleave send with blocking recv.
I'm not sure if this is better/more readable than one with multiple procs.

Reply:
ah, sorry for the missunderstanding.

What I meant by interleaving was more something like this:
```
let sent_packets: u8[4] = [0xA, 0xA, 0xA, 0xB];
let tok = for ((i, p), (tok)) in enumerate(sent_packets) {
   let tok = send(tok, enc_input_s, p);
   tok
}(tok);
let expected_packets (u8, u32)[2] = [(0xA, 3), (0xB, 1)];
let tok = for ((i, p), (tok)) in enumerate(expected_packets) {
   let (tok, output) = recv(tok, enc_output_r);
   assert_eq(output.symbol, p[0]);
   assert_eq(output.count, p[1]);
   tok
}(tok);
```
would an approach like this work? The advantage I see is that you can directly correlate test input and output for each encoding "sessions".

Reply:
Also I think I would be nice if DSLX would allow for passing variable reference to test_procs, something like:
```
#[test_proc_def] 
proc Tester<A> ....
```
and
```
#[test_proc<A=some_constant_variable>]
Tester
```
This way universal Tester could be written, and run multiple times with different payload.

Reply:
I think that's could be interesting to discuss this as part of https://github.com/google/xls/issues/81, in a way a very hard constraint on quickcheck would be to have a list of data to test with.

See also what python has for https://docs.pytest.org/en/7.1.x/example/parametrize.html

Reply:
I agree, so any way I'll change RLEEncTightTester to look as you suggested and keep old test, as I'm using parts of it in other PR.

Reply:
Thanks, I find this a lot easier to follow! wdyt?